src/sentry/ratelimits/sliding_windows.py

```
@@ -0,0 +1,209 @@
+    # quotas.
+    quotas: Sequence[Quota]
+
+
+@dataclass(frozen=True)
+class GrantedQuota:
+    # The prefix from RequestedQuota
+    prefix: str
+
+    # How much of RequestedQuota.requested can actually be used.
+    granted: int
+
+
+class SlidingWindowRateLimiter(Service):
+    def __init__(self, **options: Any) -> None:
+        pass
+
+    def check_within_quotas(
+        self, requests: Sequence[RequestedQuota], timestamp: Optional[int] = None
+    ) -> Tuple[int, Sequence[GrantedQuota]]:
+        """
+        Given a set of quotas requests and limits, compute how much quota could
+        be consumed.
+        """
+        raise NotImplementedError()
+
+    def use_quotas(
+        self, requests: Sequence[RequestedQuota], grants: Sequence[GrantedQuota], timestamp: int
+    ) -> None:
+        """
+        Given a set of requests and the corresponding return values from
+        `check_within_quotas`, consume the quotas.
+
+        This is a separate method because 1) the redis backend is not strongly
+        consistent anyway 2) the metrics string indexer may not want to
+        increment quotas if it crashes halfway through the database write. In
+        other words, it may want to first check quotas, execute the writes,
+        then apply the quotas.
+        """
+        raise NotImplementedError()
+
+    def check_and_use_quotas(
+        self, requests: Sequence[RequestedQuota], timestamp: Optional[int] = None
+    ) -> Sequence[GrantedQuota]:
+        """
+        Check the quota requests in Redis and consume the quota in one go.
+        """
+        timestamp, grants = self.check_within_quotas(requests, timestamp)
+        self.use_quotas(requests, grants, timestamp)
+        return grants
+
+
+class RedisSlidingWindowRateLimiter(SlidingWindowRateLimiter):
+    def __init__(self, **options: Any) -> None:
+        cluster_key = options.get("cluster", "default")
+        self.client = redis.redis_clusters.get(cluster_key)
```

How do we deal with redis failovers in the sentry code base? If we keep a reference to the cluster as an attribute and redis has a failover, do we let this fail and then re-establish the connection ?

---

I had to check because I don't know, but my impression is that we deal with it poorly.

Apparently we have the [RetryingRedisCluster](https://github.com/getsentry/sentry/blob/834adc3c1dc5ac1b2035ec61236e0716792f5c6d/src/sentry/utils/redis.py#L70) which reestablishes the connections transparently on some errors, but i don't know if that is enough.

We also recently added [FailoverRedis](https://github.com/getsentry/sentry/blob/834adc3c1dc5ac1b2035ec61236e0716792f5c6d/src/sentry/utils/redis.py#L92) which indicates that previously we just didn't deal with failovers? This class is only used in a single redis cluster (configured in getsentry) @beezz do you know more about this?